libopae/src/mmio.c

```diff
@@ -128,16 +130,96 @@ fpga_result __FPGA_API__ fpgaReadMMIO32(fpga_handle handle,
 	return result;
 }
 
-fpga_result __FPGA_API__ fpgaWriteMMIO64(fpga_handle handle,
+static fpga_result map_mmio_region(fpga_handle handle, uint32_t mmio_num)
+{
+	struct _fpga_handle *_handle = (struct _fpga_handle *)handle;
+	void     *addr  = NULL;
+	uint64_t wsid   = 0;
+	uint32_t flags  = 0;
+	uint64_t size   = 0;
+	uint64_t offset = 0;
+	fpga_result result = FPGA_OK;
+
+	/* Obtain MMIO region information */
+	result = port_get_region_info(handle,
+				      mmio_num,
+				      &flags,
+				      &size,
+				      &offset);
+
+	if (flags != AFU_PERMISSION) {
+		FPGA_MSG("Invalid MMIO permission flags");
+		result = FPGA_NO_ACCESS;
+		return result;
+	}
+
+	/* Map UAFU MMIO */
+	result = port_mmap_region(handle,
+				    (void **) &addr,
+				    size,
+				    PROT_READ | PROT_WRITE,
+				    offset,
+				    mmio_num);
+	if (result != FPGA_OK)
+		return result;
+
+	/* Add to MMIO list */
+	wsid = wsid_gen();
+	if (!wsid_add(&_handle->mmio_root,
+		      wsid,
+		      (uint64_t) NULL,
+		      (uint64_t) NULL,
+		      size,
+		      (uint64_t) addr,
+		      mmio_num,
+		      0)) {
+		if (munmap(addr, size)) {
+			FPGA_MSG("munmap failed. Error value is : %s",
+				 strerror(errno));
+			return FPGA_INVALID_PARAM;
+		} else {
+			FPGA_MSG("Failed to add MMIO id: %d", mmio_num);
+			return FPGA_NO_MEMORY;
+		}
+	}
+
+	return FPGA_OK;
+}
+
+/* Lazy mapping of MMIO region (only map if not already mapped) */
+static fpga_result find_or_map_wm(fpga_handle handle, uint32_t mmio_num,
+				struct wsid_map **wm_out)
+{
```

Comment:
This function will either return a valid wm pointer for the `mmio_num` supplied, or try to map the respective MMIO region and return the pointer then.